File: MergeSort.java
public class MergeSort {
    
    int[] array;

    public MergeSort(int[] array) {
        this.array = array;
    }

    public void mergeSort() {
        mergeSort(0, array.length - 1);
    }
    public void mergeSort(int l, int r) {
        if (l >= r) {
            return;
        }
        int mid = l + (r - l) / 2;
        mergeSort(l, mid);
        mergeSort(mid + 1, r);
        merge(l, mid, r);
    }
    public void merge(int l, int mid, int r) {
        int sizeOfLeft = mid - l + 1;
        int sizeOfRight = r - mid;

        int[] L = new int[sizeOfLeft];
        int[] R = new int[sizeOfRight];
        for (int i = l; i <= mid; i++) {
            L[i - l] = array[i];
        }
        for (int i = mid + 1; i <= r; i++) {
            R[i - mid - 1] = array[i];
        }

        int i = 0;
        int j = 0;
        int k = l;

        while (i < sizeOfLeft && j < sizeOfRight) {
            if (L[i] <= R[j]) {
                array[k] = L[i];
                i++;
            }
            if (R[j] < L[i]) {
                array[k] = R[j];
                j++;
            }
            k++;
        }

        while (i < sizeOfLeft) {
            array[k] = L[i];
            i++;
            k++;
        }
        while (j < sizeOfRight) {
            array[k] = R[j];
            j++;
            k++;
        }

    }




}

File: MergeSort.py
class MergeSort:

    def __init__(self, array):
        self.array = array
    def merge(self, left, mid, right):
        left_subarray =self.array[left:mid+1]
        right_subarray =self.array[mid+1:right+1]


        i = 0
        j = 0
        k = left
        while i < len(left_subarray) or j < len(right_subarray):
            if i == len(left_subarray):
                self.array[k] = right_subarray[j]
                j += 1
            elif j == len(right_subarray):
                self.array[k] = left_subarray[i]
                i += 1
            elif right_subarray[j] < left_subarray[i]:
                self.array[k] = right_subarray[j]
                j += 1
            elif left_subarray[i] <= right_subarray[j]:
                self.array[k] = left_subarray[i]
                i += 1
            k += 1

    def mergeSort(self, left=0, right=4):
        if (left >= right):
            return
        mid = left + (right - left) // 2
        self.mergeSort(left, mid)
        self.mergeSort(mid + 1, right)

        self.merge(left, mid, right)

    def print_array(self):
        print(self.array)





